plugins/hls-code-range-plugin/src/Ide/Plugin/CodeRange.hs

```diff
@@ -126,6 +150,34 @@ findPosition pos root = go Nothing root
             startOfRight <- _start . _codeRange_range <$> V.headM right
             if pos < startOfRight then binarySearchPos left else binarySearchPos right
 
+-- | Traverses through the code range and children to a folding ranges
+findFoldingRanges :: CodeRange -> [FoldingRange]
+findFoldingRanges r@(CodeRange _ children _) =
+  let frRoot :: [FoldingRange] = case createFoldingRange r of
+        Just x  -> [x]
+        Nothing -> []
+
+      frChildren :: [FoldingRange] = concat $ V.toList $ fmap findFoldingRanges children
+   in frRoot ++ frChildren
```

Comment:
Worth a note that this isn't risky because `frRoot` has  length 0 or 1, otherwise it could be quadratic. Have to be careful with `++`. You could make this totally unambiguous by pushing this inside the `frRoot` bit, i.e.
```
let frChildren = ...
in case createFoldingRange r of
  Just x -> x:frChildren
  Nothing -> frChildren
```
not sure if that's better

Reply:
Okay, This makes sense, I had not thought about this

Reply:
It's a little tricky, but `++` is linear in the first argument, so any time you see `++` and a tree traversal you want to be careful!